add DEVELOPMENT.md, update README

okartau opened the pull request:
This file explains development-related aspects,
and we move some pieces from README over there.
New material is API explanation: sockets, arguments, etc.

This resolves #179, resolves #180

okartau commented:
Does it make sense to move content of current README-qemu-notes.md to become a new chapter in DEVELOPMENT.md ?

pohly commented:
> Does it make sense to move content of current README-qemu-notes.md to become a new chapter in DEVELOPMENT.md?

Do normal developers still need that information? With "normal" I mean someone who checks out the code today, wants to build from source, make some change, and then try out that change?

Probably not, because "make start" takes care of configuring QEMU, which makes that document mostly background information that most developers can ignore. I would keep it where it is. A comment at the top could clarify that this is not something that needs to be done manually anymore.

okartau commented:
> I would keep it where it is. A comment at the top could clarify that this is not something that needs to be done manually anymore.

makes sense. I changed README-qemu-notes.md to become a new chapter in DEVELOPMENT.md? a bit and made this change as part of commit here

okartau commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,220 @@
+Network ports
+-------------
+
+Network ports are opened for serving CSI API, as configured in manifest:
```
I made it more clear by stating per port for what it's used, and not stating CSI API in heading. I also removed the "is in LISTEM state" part as it does not really add value

okartau commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,220 @@
+unix:///var/lib/kubelet/plugins/pmem-csi/csi-controller.sock
+
+
+Command line arguments
```
The idea was to describe sockets, ports, API, and command line in one document which can be treated as 'API doc'. It's true that arguments can be considered belonging to README as well. I agree, lets have them here for now. We can add note in README "arguments documented in detail in DEVELOPMENT.md"

okartau commented:
> [#181 (review)](https://github.com/intel/pmem-CSI/pull/181#pullrequestreview-208525001) still needs to be addressed.

yes, these were broken already before. Fixed now

okartau commented:
> > [#181 (review)](https://github.com/intel/pmem-CSI/pull/181#pullrequestreview-208525001) still needs to be addressed.
> 
> yes, these were broken already before. Fixed now

actually, not entirely, as "run as" link points nowhere because I had deleted that sub-heading. Needs more changes, plus addressing other comments in that review request

okartau commented:
I tried to bring the text to consistent state now, we should not discuss deploy-by-hand steps under 'make start' any more, so I moved words around.
One sentence caught my eye: `Once done, make stop will clean up the cluster and shut everything down.`
I reworded it a bit to be more focused ("once done" and "everything" are not focused  terms),
but also semantics: I think after 'make stop' and 'make start', I usually get my cluster back with pods running, so the "clean up the cluster" is not right term?

pohly commented:
True, `make stop` just shuts down in pmem-csi. It did a bit more initially in OIM.

okartau commented:
updated that part.
What about explaining ssh cmd wrappers to run commands on VM nodes, or to start ssh session there. use of ssh to nodes may not be after 'make start' ?

pohly commented:
> What about explaining ssh cmd wrappers to run commands on VM nodes, or to start ssh session there.

Yes, that's useful to have in the documentation.

> use of ssh to nodes may not be after 'make start' ?

You mean it doesn't work before `make start`? Yes, that's correct, because the VMs aren't running.

One more thing: you also add paragraphs that are just one long line. Can you wrap them?

okartau commented:
that last half-sentence was likely some leftover from moving words around. I removed some empty lines to keep paragraphs together, and added two ssh wrapper examples, please check

okartau commented:
as small surprise, this remained mergable after csi-1.0 merge.
But I anyway rebased and force pushed, adding a line to commit msg that README and README-qemu-notes were modified.

okartau commented:
There are few left-overs in README after recent csi-1.0 merge which still mention 1.12 deploy files. I think we should only cover 1.13 there. I will correct in this PR here.

Also, bit of style question, if we do not pipe deploy file through a filter, I propose lets use notion:
`deploy -f deploy/FILE`
instead of
`cat deploy/FILE | deploy -f  -`

There are notions like the latter in deploy examples in test/start-stop.make.

Also, even smaller style thing:
let's use `deploy -f deploy/FILE` instead of `deploy -f - <deploy/FILE`
as it's not clear what is advantage of the latter, or is there ?

okartau commented:
Should we delete this sentence now in README? :
`Clusters with multiple nodes with persistent memory are not fully supported at the moment. Support for this will be added when making the CSI driver topology- aware.`

okartau commented:
what is our stand now about patching the registry-address in deploy files?
`sed` example has been removed from most of places but is still present on 2 lines (the same lines I need to remove 1.12 notion). Should we still keep/give an example how to patch registry-address?

avalluri commented:
> Should we delete this sentence now in README? :
> `Clusters with multiple nodes with persistent memory are not fully supported at the moment. Support for this will be added when making the CSI driver topology- aware.`

I feel yes, we can remove this line.

avalluri commented:
> Also, even smaller style thing:
> let's use `deploy -f deploy/FILE`

We cannot use this as those files located on host, and where kubelet cannot access them over ssh. so we should use:
> deploy -f - <deploy/FILE

avalluri commented:
If this document is also targeted for cluster admins, i feel this is the right place to mention about "volume attributes/storageclass parameters" which driver understands.

okartau commented:
> We cannot use this as those files located on host, and where kubelet cannot access them over ssh

true. I realized I questioned those examples after I had adopted style of exporting KUBECONFIG and using local kubectl, which makes files accessible directly. But if one uses ssh wrappers, then indeed file needs to be given from stdin. So it depends on user's selection of kubectl run method.

avalluri commented:
> true. I realized I questioned those examples after I had adopted style of exporting KUBECONFIG and using local kubectl

oops, i missed that :) if using host kubectl then you are right.

okartau commented:
pushed another new state addressing remaining known issues:
- removed sentence about lack of multi-node cluster support
- replacing remaining 1.12 with <kubernetes version>

The "sed" minor inconsistency still remains: first creation commands have sed-example but others do not have.

okartau commented:
About documenting registry address replacement and example:
What about adding example command to replace all registry addresses under deploy?
I am using that in daily work and it frees me from thinking about the address in every creation.

`for f in $(find ./deploy -type f); do sed -i -e 's/192.168.8.1:5000/MY_REGISTRY/' $f; done`

okartau commented:
added table of contents to DEVELOPMENT, and chapter to end discussing changing DeviceMode
